src/audio/mfcc/mfcc_hifi4.c

```diff
@@ -0,0 +1,549 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2023 Intel Corporation. All rights reserved.
+//
+// Author: Andrula Song <andrula.song@intel.com>
+
+#include <sof/audio/mfcc/mfcc_comp.h>
+
+#ifdef MFCC_HIFI4
+
+#include <sof/audio/component.h>
+#include <sof/audio/audio_stream.h>
+#include <sof/math/auditory.h>
+#include <sof/math/matrix.h>
+#include <sof/math/sqrt.h>
+#include <sof/math/trig.h>
+#include <sof/math/window.h>
+#include <sof/trace/trace.h>
+#include <user/mfcc.h>
+#include <errno.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <xtensa/tie/xt_hifi4.h>
+
+/* MFCC with 16 bit FFT benefits from data normalize, for 32 bits there's no
+ * significant impact. The amount of left shifts for FFT input is limited to
+ * 10 that equals about 60 dB boost. The boost is compensated in Mel energy
+ * calculation.
+ */
+#if MFCC_FFT_BITS == 16
+#define MFCC_NORMALIZE_FFT
+#else
+#undef MFCC_NORMALIZE_FFT
+#endif
+#define MFCC_NORMALIZE_MAX_SHIFT	10
+
+/* Open files for data debug output in testbench */
+#define DEBUG_WITH_TESTBENCH
+#if defined(CONFIG_LIBRARY) && defined(DEBUG_WITH_TESTBENCH)
+#include <stdio.h>
+#define DEBUGFILES
+#undef DEBUGFILES_READ_FFT /* Override FFT out with file input */
+#undef DEBUGFILES_READ_MEL /* Override Mel filterbank with file input */
+#endif
+
+#ifdef MFCC_DEBUGFILES
+#ifdef DEBUGFILES_READ_FFT
+FILE *fh_fft_out_read;
+#endif
+#ifdef DEBUGFILES_READ_MEL
+FILE *fh_mel_read;
+#endif
+FILE *fh_fft_in;
+FILE *fh_fft_out;
+FILE *fh_pow;
+FILE *fh_mel;
+FILE *fh_ceps;
+
+void mfcc_generic_debug_open(void)
+{
+#ifdef DEBUGFILES_READ_FFT
+	fh_fft_out_read = fopen("ref_fft_out.txt", "r");
+#endif
+#ifdef DEBUGFILES_READ_MEL
+	fh_mel_read = fopen("ref_fft_mel.txt", "r");
+#endif
+	fh_fft_in = fopen("fft_in.txt", "w");
+	fh_fft_out = fopen("fft_out.txt", "w");
+	fh_pow = fopen("fft_pow.txt", "w");
+	fh_mel = fopen("fft_mel.txt", "w");
+	fh_ceps = fopen("ceps.txt", "w");
+}
+
+void mfcc_generic_debug_close(void)
+{
+#ifdef DEBUGFILES_READ_FFT
+	fclose(fh_fft_out_read);
+#endif
+#ifdef DEBUGFILES_READ_MEL
+	fclose(fh_mel_read);
+#endif
+	fclose(fh_fft_in);
```

Comment:
Where does fclose() come from ? Is this only used for testing on the host ? Reason I ask is that this will collide with Zephyr C library.

Reply:
yeah, it is only for testing on the host. Actually I did not finish all function level optimization, the matrix functions can be optimized too. I only use the simulator statistics to make sure the output of the function is the same, but can not see the performance of the whole algorithm, so keep the file testing to sync with Seppo after he come back.

Reply:
ok, we need to remove the posix file APIs here since primamry module target is DSP. Lets wait for Seppo.